Support AccessMode.NONE

craigtaverner opened the pull request:
Support the concept of an AccessMode.NONE allowing transactions to be created, but no actual operations to be performed. This is required in order to allow admin activities through Cypher that do not touch the database, like password change.

pgpv committed:
prototype write only mode

craigtaverner committed:
Symmetric exception behavior for read, write and schema access

craigtaverner commented:
This replaces the PR with common commits at https://github.com/neo4j/neo4j/pull/6462

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -74,34 +75,52 @@
                 };
 
-        TransactionType upgradeToDataTransaction() throws InvalidTransactionTypeKernelException
+        TransactionType enableReadTransaction() throws IllegalStateException
```
This does not throw IllegalStateException

henriknyman commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/KernelTransactionAccessModeTest.java:
```diff
@@ -36,6 +37,20 @@
     @Rule public ExpectedException exception = ExpectedException.none();
 
     @Test
```
We should add a test for not allowing reads when AccessMode.NONE

henriknyman commented on community/server/src/main/java/org/neo4j/server/rest/transactional/TransactionHandle.java:
```diff
@@ -395,7 +396,7 @@ private void executePeriodicCommitStatement(
+            catch ( KernelException | CypherException | AuthorizationViolationException e )
             {
                 errors.add( new Neo4jError( e.status(), e ) );
             }
```
I think we need to catch AuthorizationViolationException in AuthorizationFilter.doFilter to cover all REST endpoints, but it could also be done in a separate PR.

henriknyman committed:
Fix review comments

- Add tests for AccessMode.NONE
- Add unauthorized access violation error for REST in AuthorizationFilter

craigtaverner commented:
I realised that the coupling between the transaction type and the accessmode is not necessary or even useful. I think we can do this better by reverting the transaction type to previous behavior and asserting accessmode in the statement.

craigtaverner committed:
Simplify, by decoupling TransactionType from AccessMode

craigtaverner committed:
Fix failing test due to message text change

craigtaverner committed:
Fixed failing tests due to missing mocks for AccessMode

henriknyman committed:
Add integration test for REST endpoints

The test is disabled until we have authorization in place

OliviaYtterbrink commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;wJ2xxLLY&#x2F;42-support-accessmode-none"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Support AccessMode.NONE</a>

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/AccessMode.java:
```diff
@@ -66,7 +66,29 @@ public boolean allowsSchemaWrites()
 
-    /** Allows reading data and schema, but not writing. */
+    /** Allows writing data */
+    WRITE_ONLY
```
why this access mode was introduced here?/ Should it be part of this PR?

craigtaverner replied:
This code is not strictly required by the current version, but will be needed in future, so we thought it simple enough, and small enough, to be OK to include now. If you have a strong feeling about this, we can remove it and add it back later.

MishaDemianenko replied:
sure fine

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -131,7 +131,7 @@ TransactionType upgradeToSchemaTransaction() throws InvalidTransactionTypeKernel
     private TransactionHooks.TransactionHooksState hooksState;
     private KernelStatement currentStatement;
```
`KernelTransactionImplementation` still contains `TransactionType` enumeration and one of possible state is READ which intersects with AccessMode values and actual responsibility.

craigtaverner replied:
These two have different concerns, enforce different things and throw different exceptions on failure. We did make the mistake in earlier versions of this PR of bringing these two closer together, but that was confusing the situation, so we've corrected that by separating them again. The TransactionType is now identical to what it was before, which is an internal transaction state tracking whether or not a transaction contains write or schema commands (and enforcing that it cannot contain both). This is entirely orthogonal to the authorisation concern of what a transaction is allowed to do. To keep this separation clear, we moved the AccessMode checks to the KernelStatement. The KernelTransactionImplementation is merely a carrier of the AccessMode. Enforcement of what is allowed is done in KernelStatement.

MishaDemianenko replied:
fully agree with all your statements, but still TransactionType contains `READ` which should not be there anymore because we have `AccessMode` now right?

craigtaverner replied:
I see what you mean. I double-checked and the original value was 'ANY'. We changed it to 'READ' because it better represented the state of the transaction. The TransactionType is supposed to track when the transaction upgrades in one of two directions, 'data write' and 'schema write'. So the question is what state was it in before this 'upgrade'. We perceived it to be in a 'read' state, meaning that it was not yet performing any kind of write operation. After the upgrade it would be in one of two possible 'write' states.

The older code uses 'ANY' to mark a transaction containing only READ commands, before upgrading to one of the two WRITE states. I think the use of READ makes more sense than ANY. I also think this choice is not related at all to the AccessMode, and makes sense for the TransactionType itself.

MishaDemianenko replied:
From your description i can conclude that TransactionType represent currently what kind of writes where performed in transaction. In this case it sounds reasonable to me to call initial state NONE, rename TransactionState to something that will actually name correctly what was just described and it should be javadoced in the code

MishaDemianenko commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/security/AuthorizationViolationException.java:
```diff
@@ -0,0 +1,49 @@
+        super( msg );
+    }
+
+    public AuthorizationViolationException( String msg, Throwable cause )
```
constructor not used

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java:
```diff
@@ -58,6 +59,11 @@ public KernelStatement( KernelTransactionImplementation transaction,
     @Override
     public ReadOperations readOperations()
     {
+        if( !transaction.mode().allowsReads() )
```
personally i would prefer to have a bit different API, like `transaction.mode().verifyAllowReads()` which will throw exception in case if current mode do not allow reads; that will allow hide all those exceptions and check login inside of access mode and will keep KernelStatement more clear.

craigtaverner replied:
That is an option. The 'if' statement would occur elsewhere, so this is just moving code around a bit. We went for this approach here partly because it was more similar to where we think things will go in future when we write richer access control into the statement. Right now there is a 1:1:1 mapping from the kernel statement methods, access mode and exceptions thrown, so either solution is fine, but in future we think it might be more complex. For example, the AccessMode might not have enough knowledge to compose the message in the exception, and that would need to be done in the statement. I'd prefer to keep things the way they are here.

MishaDemianenko replied:
ok fine :)

MishaDemianenko commented:
@craigtaverner review round completed

MishaDemianenko commented:
@petraselmer @craigtaverner why this was merged?
